src/Sylius/Bundle/CoreBundle/CatalogPromotion/CommandHandler/RemoveCatalogPromotionHandler.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\CatalogPromotion\CommandHandler;
+
+use Doctrine\ORM\EntityManager;
+use Sylius\Bundle\CoreBundle\CatalogPromotion\Command\RemoveCatalogPromotion;
+use Sylius\Component\Core\Model\CatalogPromotionInterface;
+use Sylius\Component\Promotion\Exception\InvalidCatalogPromotionStateException;
+use Sylius\Component\Promotion\Model\CatalogPromotionStates;
+use Sylius\Component\Promotion\Repository\CatalogPromotionRepositoryInterface;
+
+final class RemoveCatalogPromotionHandler
+{
+    public function __construct(
+        private CatalogPromotionRepositoryInterface $catalogPromotionRepository,
+        private EntityManager $entityManager,
+    ) {
+    }
+
+    public function __invoke(RemoveCatalogPromotion $command): void
+    {
+        /** @var CatalogPromotionInterface|null $catalogPromotion */
+        $catalogPromotion = $this->catalogPromotionRepository->findOneBy(['code' => $command->code]);
+
+        if (null === $catalogPromotion) {
+            return;
+        }
+
+        if ($catalogPromotion->getState() !== CatalogPromotionStates::STATE_PROCESSING) {
```

Comment:
if we now remove a catalog promotion that is in the `processing` state only, are we confident that this catalog promotion has already been removed from all products at this point?

Reply:
Unfortunately, we can't be sure catalog promotion has been removed from all products. But nothing has changed here. Before, with the inactive state, we also couldn't be confident about that. However, if the CP is active on some product variant, the deletion will fail with constraint violation from the database. Then there are 3 retries before the message is put to the failure transport.

The @lchrusciel's proposal https://github.com/Sylius/Sylius/pull/14568#pullrequestreview-1194181643 will deal with it so thanks to this we will be able to even track how many batches have left and when we'll reach zero baches we can dispatch `RemoveCatalogPromotion` event. I created a task on our Jira to do so :+1: